nni/retiarii/hub/pytorch/autoformer.py

```diff
@@ -0,0 +1,464 @@
+    random_tensor.floor_()  # binarize
+    output = x.div(keep_prob) * random_tensor
+    return output
+
+
+class DropPath(nn.Module):
+    """Drop paths (Stochastic Depth) per sample  (when applied in main path of residual blocks).
+    """
+
+    def __init__(self, drop_prob=None):
+        super(DropPath, self).__init__()
+        self.drop_prob = drop_prob
+
+    def forward(self, x):
+        return drop_path(x, self.drop_prob, self.training)
+
+
+class RelativePosition2D(nn.Module):
+    def __init__(self, head_embed_dim, length=14,) -> None:
+        super().__init__()
+        self.head_embed_dim = head_embed_dim
+        self.legnth = length
+        self.embeddings_table_v = nn.Parameter(
+            torch.randn(length * 2 + 2, head_embed_dim))
+        self.embeddings_table_h = nn.Parameter(
+            torch.randn(length * 2 + 2, head_embed_dim))
+
+        trunc_normal_(self.embeddings_table_v, std=.02)
+        trunc_normal_(self.embeddings_table_h, std=.02)
+
+    def forward(self, length_q, length_k):
+        # remove the first cls token distance computation
+        length_q = length_q - 1
+        length_k = length_k - 1
+        range_vec_q = torch.arange(length_q)
+        range_vec_k = torch.arange(length_k)
+        # compute the row and column distance
+        distance_mat_v = (range_vec_k[None, :] //
+                          int(length_q ** 0.5) -
+                          range_vec_q[:, None] //
+                          int(length_q ** 0.5))
+        distance_mat_h = (range_vec_k[None, :] %
+                          int(length_q ** 0.5) -
+                          range_vec_q[:, None] %
+                          int(length_q ** 0.5))
+        # clip the distance to the range of [-legnth, legnth]
+        distance_mat_clipped_v = torch.clamp(
+            distance_mat_v, -self.legnth, self.legnth)
+        distance_mat_clipped_h = torch.clamp(
+            distance_mat_h, -self.legnth, self.legnth)
+
+        # translate the distance from [1, 2 * legnth + 1], 0 is for the cls
+        # token
+        final_mat_v = distance_mat_clipped_v + self.legnth + 1
+        final_mat_h = distance_mat_clipped_h + self.legnth + 1
+        # pad the 0 which represent the cls token
+        final_mat_v = F.pad(
+            final_mat_v, (1, 0, 1, 0), "constant", 0)
+        final_mat_h = F.pad(
+            final_mat_h, (1, 0, 1, 0), "constant", 0)
+
+        final_mat_v = torch.tensor(
+            final_mat_v,
+            dtype=torch.long,
+            device=self.embeddings_table_v.device)
+        final_mat_h = torch.tensor(
+            final_mat_h,
+            dtype=torch.long,
+            device=self.embeddings_table_v.device)
+        # get the embeddings with the corresponding distance
+        embeddings = self.embeddings_table_v[final_mat_v] + \
+            self.embeddings_table_h[final_mat_h]
+
+        return embeddings
+
+
+class Attention(nn.Module):
+    def __init__(
```

Comment:
Adding a reference of this "Attention" here.
I suggest naming it "RelativePositionAttention".